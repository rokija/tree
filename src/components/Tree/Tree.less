.Tree {
    display: flex;
    justify-content: space-between;
}
