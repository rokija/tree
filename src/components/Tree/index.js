import React, { Component, Fragment } from 'react'
import Leaf from '../Leaf'

class Tree extends Component {
    state = {
      inputValue: '',
      selectedNodeIndex: null,
      children: [{
        id: 1,
        parentId: 0,
        value: 'root leaf',
        paddingLeft: 10,
      }],
    }

    onInputChange = e => this.setState({ inputValue: e.target.value })

    onOpenInput = index => this.setState({ selectedNodeIndex: index })

    onAddNode = (id) => {
      const { children, inputValue } = this.state
      const parentElement = children.find(child => child.id === id)

      this.setState({
        children: [...children, {
          id: id + 1, parentId: id, value: inputValue, paddingLeft: parentElement.paddingLeft + 10,
        }],
        inputValue: '',
        selectedNodeIndex: null,
      })
    }

    render() {
      const { children, selectedNodeIndex } = this.state

      return (
        <div>
          {children.map((leaf, i) => (
            <div key={`leaf-${i}`}>
              <Leaf
                index={i}
                paddingLeft={leaf.paddingLeft}
                value={leaf.value}
                onOpenInput={this.onOpenInput}
              />
              {selectedNodeIndex === i && (
                <Fragment>
                  <input onChange={this.onInputChange} />
                  <button onClick={() => this.onAddNode(leaf.id)}>+</button>
                </Fragment>
              )}
            </div>
          ))}
        </div>
      )
    }
}

export default Tree
