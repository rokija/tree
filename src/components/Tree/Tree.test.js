import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import Tree from '.'

describe('Tree', () => {
  const wrapper = shallow(<Tree />)
  const wrapperInstance = wrapper.instance()

  it('renders properly', () => {
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  it('sets value in state when calling onInputChange', () => {
    wrapperInstance.onInputChange({ target: { value: 'new value' } })
    expect(wrapper.state().inputValue).toEqual('new value')
  })

  it('calls onOpenInput and sets provided index in state', () => {
    wrapperInstance.onOpenInput(4)
    expect(wrapper.state().selectedNodeIndex).toEqual(4)
  })

  it('calls onAddNode and adds a new object in children array in state', () => {
    const defaultState = {
      children: [{
        id: 1,
        parentId: 0,
        value: 'root leaf',
        paddingLeft: 10,
      }],
    }

    wrapperInstance.onAddNode(1)
    expect(wrapper.state().children).toEqual([...defaultState.children, {
      id: 2, paddingLeft: 20, parentId: 1, value: 'new value',
    }])
  })
})
