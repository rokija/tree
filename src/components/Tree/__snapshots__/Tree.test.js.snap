// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tree renders properly 1`] = `
<div>
  <div
    key="leaf-0"
  >
    <Leaf
      index={0}
      onOpenInput={[Function]}
      paddingLeft={10}
      value="root leaf"
    />
  </div>
</div>
`;
